tests/helpers/test_entity_platform.py

```diff
@@ -560,22 +561,28 @@ async def test_async_remove_with_platform_update_finishes(hass: HomeAssistant) -
     entity1 = MockEntity(name="test_1")
 
     async def _delayed_update(*args, **kwargs):
-        await asyncio.sleep(0.01)
+        update_called.set()
+        await update_done.wait()
 
     entity1.async_update = _delayed_update
 
-    # Add, remove, add, remove and make sure no updates
+    # Add, remove, and make sure no updates
     # cause the entity to reappear after removal
     for _ in range(2):
+        update_called = asyncio.Event()
+        update_done = asyncio.Event()
         await component.async_add_entities([entity1])
         assert len(hass.states.async_entity_ids()) == 1
         entity1.async_write_ha_state()
         assert hass.states.get(entity1.entity_id) is not None
         task = asyncio.create_task(entity1.async_update_ha_state(True))
+        await update_called.wait()
         await entity1.async_remove()
         assert len(hass.states.async_entity_ids()) == 0
+        update_done.set()
         await task
         assert len(hass.states.async_entity_ids()) == 0
+        entity1._platform_state = EntityPlatformState.NOT_ADDED
```

Comment:
Why do we change the internal state of the entity? Isn't that what we should test that the entity handles correctly in its lifecycle?

Reply:
I agree the way I changed the test is not good.
@bdraco why did you make the test do the remove + late update sing and dance repeat twice? If there's a good reason we could change the entity's entity id a couple of times instead.

Reply:
If I remember correctly, the goal was to detect a race.  I'll have to do some git archeology later today to find the original PR to better remember the reasoning.

Reply:
I'm pretty sure the problem was the entity was updating after it was removed and than the state was added to the state machine which caused the entity id to be unable to be reused after a reload since it way already in use

Reply:
That makes sense, but then the loop should be unnecessary?

Reply:
probably doesn't need to be a loop, just needs to make sure it can't write to the state machine when its not added

Reply:
OK, I rewrote the test so it adds another entity object to make sure we get the same entity id again